reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -0,0 +1,646 @@
+            return;
+        }
+        source.subscribe(new SwitchOnFirstInner<>(actual, transformer));
+    }
+
+    static final class SwitchOnFirstInner<T, R> extends Flux<T>
+            implements InnerOperator<T, R> {
+
+        final CoreSubscriber<? super R>                                        outer;
+        final BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer;
+
+        Subscription s;
+        Throwable    throwable;
+
+        volatile T       first;
+        volatile boolean done;
+        volatile boolean cancelled;
+
+        volatile CoreSubscriber<? super T> inner;
+        @SuppressWarnings("rawtypes")
+        static final AtomicReferenceFieldUpdater<SwitchOnFirstInner, CoreSubscriber> INNER =
+                AtomicReferenceFieldUpdater.newUpdater(SwitchOnFirstInner.class, CoreSubscriber.class, "inner");
+
+        volatile int wip;
+        @SuppressWarnings("rawtypes")
+        static final AtomicIntegerFieldUpdater<SwitchOnFirstInner> WIP =
+                AtomicIntegerFieldUpdater.newUpdater(SwitchOnFirstInner.class, "wip");
+
+        volatile int once;
+        @SuppressWarnings("rawtypes")
+        static final AtomicIntegerFieldUpdater<SwitchOnFirstInner> ONCE =
+                AtomicIntegerFieldUpdater.newUpdater(SwitchOnFirstInner.class, "once");
+
+        SwitchOnFirstInner(
+                CoreSubscriber<? super R> outer,
+                BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer) {
+            this.outer = outer;
+            this.transformer = transformer;
+        }
+
+        @Override
+        @Nullable
+        public Object scanUnsafe(Attr key) {
+            if (key == Attr.CANCELLED) return cancelled;
+            if (key == Attr.TERMINATED) return done || cancelled;
+
+            return InnerOperator.super.scanUnsafe(key);
+        }
+
+        @Override
+        public CoreSubscriber<? super R> actual() {
+            return outer;
+        }
+
+        @Override
+        public Context currentContext() {
+            CoreSubscriber<? super T> actual = inner;
+
+            if (actual != null) {
+                return actual.currentContext();
+            }
+
+            return outer.currentContext();
+        }
+
+        @Override
+        public void cancel() {
+            if (!cancelled) {
+                cancelled = true;
+                s.cancel();
+
+                if (WIP.getAndIncrement(this) == 0) {
+                    INNER.lazySet(this, null);
+
+                    T f = first;
+                    if (f != null) {
+                        first = null;
+                        Operators.onDiscard(f, currentContext());
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void subscribe(CoreSubscriber<? super T> actual) {
+            if (once == 0 && ONCE.compareAndSet(this, 0, 1)) {
```

Comment:
`ONCE` is more of a `STATE`. maybe have the states as constant would help readability here (STATE_INIT = 0, STATE_SUBSCRIBED_FOR_FIRST = 1, STATE_SUBSCRIBED_FOR_TRANSFORMED = 2 or something like that)

Reply:
fixed